acceptance/helpers.rb

```diff
@@ -0,0 +1,230 @@
+      #    - If a `site.pp` already exists there, record its contents and
+      #      permissions and create a teardown task to restore them after the test
+      #      finishes.
+      # 2. Using {Beaker::DSL::PuppetHelpers.with_puppet_running_on}:
+      #    - Perform a puppet run on the agents, and
+      #    - execute the block, if any.
+      #
+      # @see Beaker::DSL::PuppetHelpers.with_puppet_running_on
+      # @param [String] site_pp_contents The contents of the default site.pp file. This
+      #   content will be wrapped as follows, so that it applies to all nodes:
+      #     ```
+      #     node default {
+      #       #{site_pp_contents}
+      #     }
+      #     ```
+      # @param [Hash] master_opts Options to pass to {Beaker::DSL::PuppetHelpers.with_puppet_running_on}.
+      # @yield Invokes {Beaker::DSL::PuppetHelpers.with_puppet_running_on},
+      # passing along master_opts, if supplied.
+      def with_default_site_pp(site_pp_contents, master_opts = {})
+        manifest_contents = %(node default { #{site_pp_contents} })
+
+        # PMT will have installed dependencies in the production environment; We will put our manifest there, too:
+        site_pp_path = File.join(puppet_config(master, 'codedir'), 'environments', 'production', 'manifests', 'site.pp')
+
+        if file_exists_on(master, site_pp_path)
+          original_contents = file_contents_on(master, site_pp_path)
+          original_perms = on(master, %(stat -c "%a" #{site_pp_path})).stdout.strip
+
+          teardown do
+            on(master, %(echo "#{original_contents}" > #{site_pp_path}))
+            on(master, %(chmod #{original_perms} #{site_pp_path}))
+          end
+        else
+          teardown do
+            on(master, "rm -f #{site_pp_path}")
+          end
+        end
+
+        create_remote_file(master, site_pp_path, manifest_contents)
+        on(master, %(chown #{puppet_user(master)} "#{site_pp_path}"))
+        on(master, %(chmod 755 "#{site_pp_path}"))
+
+        with_puppet_running_on(master, master_opts) do
+          on(agents_only, puppet(%(agent --test --server #{master.hostname})), acceptable_exit_codes: [0, 2])
+          yield if block_given?
+        end
+      end
+
+      # Wraps {with_default_site_pp} to:
+      #
+      # - Install the a puppet-agent package (`initial_package_version_or_collection`)
+      #   on all the agent hosts,
+      # - Put a teardown step in place so that this package is uninstalled after the test,
+      # - Run a default manifest on all the agent hosts, and
+      # - Allow for assertions inside in a block
+      #
+      # @param [String] initial_package_version_or_collection Either a version
+      #   of puppet-agent or the name of a puppet collection to install on agent hosts
+      # @param [String] upgrade_manifest A manifest to apply to all agent nodes
+      def run_foss_upgrade_with_manifest(initial_package_version_or_collection, upgrade_manifest)
```

Comment:
I think the code here's short enough and that there are enough useful helpers already that it would be good enough to leave this code to the individual test files, especially when we'll start adding more complicated tests where the preparation's likely going to be more involved (like Windows service stuff).